### DISCUSSION TITLE
"combat.push_nodes" setting question

### DISCUSSION THREAD

**Discussion by Danielosama** [XVs32/kcauto_custom#62] [Q&A] [answered]
Hello,

I had a question about how the setting _combat.push_nodes_ works exactly.

I tried in the 3-2 Node B for ammo farming:
![image](https://github.com/XVs32/kcauto_custom/assets/12687427/ad6ca8c3-1b1c-4922-829e-f9dbac60d063)

As one can see, even if a ship gets Taiha at node A, you can advance to node B for the ammo wtihout risk.

This is my config:
`	"combat.push_nodes":	["A"],`

However, the script did not push, and instead retreated when a ship got Taihad at node A.

Does this setting work? If so, what node should be chosen as "pushed"?

Thank you for your time.

All best,

**Comment by XVs32** [comment, marked as answer]
Hi [Danielosama](https://github.com/Danielosama),

From my understanding, your `push_nodes` setting is just fine.

There are two things that stops the node push from working:
1. There is a logic bug in the original kcauto, and TBH, I've never touch the node push function.
Cause it looks dangerous to me.
Anyway, here is the `git diff` which could fix the problem.
```diff
diff --git a/kcauto/combat/combat_core.py b/kcauto/combat/combat_core.py
index ef5a25c..c5c3a98 100644
--- a/kcauto/combat/combat_core.py
+++ b/kcauto/combat/combat_core.py
@@ -561,30 +561,30 @@ class CombatCore(CoreBase):
         Log.log_debug("Resolve continue sortie prompt.")
         continue_sortie = True
         retreat_limit = cfg.config.combat.retreat_limit
-        if self.current_node in cfg.config.combat.push_nodes:
+        if NodeEnum(self.current_node.name) in cfg.config.combat.push_nodes:
             Log.log_msg(f"{self.current_node} is specified as a push node.")
-            return True
-        for fleet in flt.fleets.combat_fleets:
-            if fleet.weakest_state >= retreat_limit:
-                Log.log_warn(
-                    f"Fleet {fleet.fleet_id} has ships with "
-                    f"{retreat_limit.display_name} damage or above.")
+        else:
+            for fleet in flt.fleets.combat_fleets:
+                if fleet.weakest_state >= retreat_limit:
+                    Log.log_warn(
+                        f"Fleet {fleet.fleet_id} has ships with "
+                        f"{retreat_limit.display_name} damage or above.")
+                    continue_sortie = False
+                elif fleet.visual_health['heavy'] > 0:
+                    Log.log_warn(
+                        f"Fleet {fleet.fleet_id} has a critically damaged ship "
+                        "not calculated from the API.")
+                    continue_sortie = False
+            if (
+                    NodeEnum(self.current_node.name)
+                    in cfg.config.combat.retreat_points):
+                Log.log_debug("Retreat specified in config.")
                 continue_sortie = False
-            elif fleet.visual_health['heavy'] > 0:
-                Log.log_warn(
-                    f"Fleet {fleet.fleet_id} has a critically damaged ship "
-                    "not calculated from the API.")
+            if (
+                    NodeEnum(len(self.combat_nodes_run))
+                    in cfg.config.combat.retreat_points):
+                Log.log_debug("Retreat specified combat # in config.")
                 continue_sortie = False
-        if (
-                NodeEnum(self.current_node.name)
-                in cfg.config.combat.retreat_points):
-            Log.log_debug("Retreat specified in config.")
-            continue_sortie = False
-        if (
-                NodeEnum(len(self.combat_nodes_run))
-                in cfg.config.combat.retreat_points):
-            Log.log_debug("Retreat specified combat # in config.")
-            continue_sortie = False
 
         if continue_sortie:
             Log.log_msg("Continuing sortie.")

```
~2. I've added a `combat.override` option in the latest version,
you would want to enable that (`"combat.override": true,`) in order to use your own combat setting.
Otherwise, kcauto_custom would use the default combat setting in `data/config/combat`.~

A successful run would shows you log like this:
![image](https://github.com/XVs32/kcauto_custom/assets/16824564/e7c06f99-f117-4a04-8359-197ef6a9122d)

Anyway, node push is definitely a function in danger zone, 
**enable with extra cautious please**.

Sincerely,
XVs32

**Comment by Danielosama** [comment]
Thanks for the answer.

I am afraid I don't really understand this part:

```
I've added a combat.override option in the latest version,
you would want to enable that ("combat.override": true,) in order to use your own combat setting.
Otherwise, kcauto_custom would use the default combat setting in data/config/combat.
```

What's the default combat setting exactly? (can't find anything in `data/config/combat`).
Are you talking about `combat_core.py`?
Where would be my "own combat setting"?

Sorry for all these questions.

All best,

**Comment by XVs32** [comment]
Sorry for the confusion, I have double checked the release code,
apparently the default combat config is not in the V2.0.0 release pack yet.

I think it is ok for you to ignore "combat.override" for now, or you can just enable it for safe.
(Or maybe you don't even have the "combat.override" option in your config?)
Either way, just ignore it for now.

---

Default combat config is a set of config files made for each map,
for example, if you tell kcauto to sorite to 4-5, 
![image](https://github.com/XVs32/kcauto_custom/assets/16824564/3186cd31-4e31-46f9-959f-cd16ff3447a4)
we used to have to manually set 
```
"combat.node_selects": [ 
         "A>D",
        "C>D"
]
```
In the next version, kcauto_custom will load a combat config specifically made for 4-5, 
that way we don't have to manually tweak the config when sortie to maps like 4-5(node select), 7-4(LBAS) etc.


Actually it was done long ago in my local branch, 
so I thought it was released.

Sorry for the confusion.

Sincerely,
XVs32

**Reply by Danielosama** [reply]
Correct, I do not see the "combat.override" option anywhere in the config (not even in the _templates_ ones like basic-7-4.sjon)

I understand now the default combat config.

However, if I'm making my own and write a command line to load that one specifically (let's say it's called _3-2-ammo.json_), would it load a default config or load mine **even** if _combat.override_ was set to false?

**Reply by XVs32** [reply]
As long as `combat.override` is `false`, kcauto_custom will go for default config.
It doesn't matter if kcauto_custom is start from command line or cui.

**Reply by Danielosama** [reply]
Interesting, and which default config will it choose from all the different map ones?
